hack/e2e.sh

```diff
@@ -56,40 +62,234 @@ while getopts "h?" opt; do
-    export NO_BUILD=y
-fi
+# https://github.com/kubernetes-sigs/kind/releases/tag/v0.6.0
+declare -A kind_node_images
+kind_node_images["v1.11.10"]="kindest/node:v1.11.10@sha256:44e1023d3a42281c69c255958e09264b5ac787c20a7b95caf2d23f8d8f3746f2"
+kind_node_images["v1.12.10"]="kindest/node:v1.12.10@sha256:e93e70143f22856bd652f03da880bfc70902b736750f0a68e5e66d70de236e40"
+kind_node_images["v1.13.12"]="kindest/node:v1.13.12@sha256:ad1dd06aca2b85601f882ba1df4fdc03d5a57b304652d0e81476580310ba6289"
+kind_node_images["v1.14.9"]="kindest/node:v1.14.9@sha256:00fb7d424076ed07c157eedaa3dd36bc478384c6d7635c5755746f151359320f"
+kind_node_images["v1.15.6"]="kindest/node:v1.15.6@sha256:1c8ceac6e6b48ea74cecae732e6ef108bc7864d8eca8d211d6efb58d6566c40a"
+kind_node_images["v1.16.3"]="kindest/node:v1.16.3@sha256:bced4bc71380b59873ea3917afe9fb35b00e174d22f50c7cab9188eac2b0fb88"
+
+function e2e::image_build() {
+    if [ -n "$SKIP_BUILD" ]; then
+        echo "info: skip building images"
+        export NO_BUILD=y
+    fi
+    if [ -n "$SKIP_IMAGE_BUILD" ]; then
+        echo "info: skip building and pushing images"
+        return
+    fi
+    DOCKER_REGISTRY=$DOCKER_REGISTRY IMAGE_TAG=$IMAGE_TAG make docker
+    DOCKER_REGISTRY=$DOCKER_REGISTRY IMAGE_TAG=$IMAGE_TAG make e2e-docker
+    DOCKER_REGISTRY=$DOCKER_REGISTRY IMAGE_TAG=$IMAGE_TAG make test-apiesrver-docker
+}
+
+function e2e::image_load() {
+    local names=(
+        pingcap/tidb-operator
+        pingcap/tidb-operator-e2e
+        pingcap/test-apiserver
+    )
+    for n in ${names[@]}; do
+        $KIND_BIN load docker-image --name $CLUSTER $DOCKER_REGISTRY/$n:$IMAGE_TAG
+    done
+}
+
+function e2e::cluster_exists() {
+    local name="$1"
+    $KIND_BIN get clusters | grep $CLUSTER &>/dev/null
+}
+
+function e2e::up() {
+    if [ -n "$SKIP_UP" ]; then
+        echo "info: skip starting a new cluster"
+        return
+    fi
+    if e2e::cluster_exists $CLUSTER; then
+        echo "info: deleting the cluster '$CLUSTER'"
+        $KIND_BIN delete cluster --name $CLUSTER
+    fi
+    echo "info: starting a new cluster"
+    tmpfile=$(mktemp)
+    trap "test -f $tmpfile && rm $tmpfile" RETURN
+    cat <<EOF > $tmpfile
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+EOF
+    # control-plane
+    cat <<EOF >> $tmpfile
+nodes:
+- role: control-plane
+EOF
+    if [[ "$KIND_DATA_HOSTPATH" != "none" ]]; then
+        if [ ! -d "$KIND_DATA_HOSTPATH" ]; then
+            echo "error: '$KIND_DATA_HOSTPATH' is not a directory"
+            exit 1
+        fi
+        local hostWorkerPath="${KIND_DATA_HOSTPATH}/control-plane"
+        test -d $hostWorkerPath || mkdir $hostWorkerPath
+        cat <<EOF >> $tmpfile
+  extraMounts:
+  - containerPath: /mnt/disks/
+    hostPath: "$hostWorkerPath"
+    propagation: HostToContainer
+EOF
+    fi
+    # workers
+    for ((i=1; i <=3; i++)) {
+        cat <<EOF >> $tmpfile
+- role: worker
+EOF
+        if [[ "$KIND_DATA_HOSTPATH" != "none" ]]; then
```

Comment:
The script shebang is bash by default. Does the conditional judgment statement use double brackets uniformly?

Reply:
I don't think it's necessary. In the future, we can use a static analysis tool [shellcheck](https://github.com/koalaman/shellcheck) which can help us in writing robust shell scripts.

Reply:
I think there is no problem using single brackets or double brackets. The main question is whether we can use one of them in the script instead of mixing them.

Reply:
I'm fine with mixing them. `[[` is preferred over `[` in `bash` because it has some enhancements, however, `[` / `test` commands are often shorter in simple test cases.

Reply:
we can stick to `[[` when we have some verify scripts to force users not to use `[`, otherwise it's a burden to check it case by case